compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp

```diff
@@ -0,0 +1,460 @@
+
+  // Check that the producer of the add is an arith.mul op
+  Operation *mulOp;
+  if (add0.isa<BlockArgument>()) {
+    mulOp = add1.getDefiningOp();
+  } else {
+    mulOp = add0.getDefiningOp();
+  }
+  if (!mulOp || mulOp->getNumOperands() != 2) {
+    LLVM_DEBUG(llvm::dbgs() << "no arith.mul body op, wrong numOperands\n");
+    return false;
+  }
+  if (!matchPattern(mulOp, m_Op<arith::MulFOp>()) &&
+      !matchPattern(mulOp, m_Op<arith::MulIOp>())) {
+    LLVM_DEBUG(llvm::dbgs() << "no arith.mul body op\n");
+    return false;
+  }
+  Value mul0 = mulOp->getOperand(0);
+  Value mul1 = mulOp->getOperand(1);
+  if (mul0.isa<BlockArgument>() && mul1.isa<BlockArgument>()) {
+    return true;
+  }
+
+  // Check that non block args come from arith.ext ops
+  if (!mul0.isa<BlockArgument>()) {
+    Operation *extOp0 = mul0.getDefiningOp();
+    if (!extOp0 || extOp0->getNumOperands() != 1) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "no arith.ext body op on in0, wrong numOperands\n");
+      return false;
+    }
+    if (!matchPattern(extOp0, m_Op<arith::ExtUIOp>()) &&
+        !matchPattern(extOp0, m_Op<arith::ExtSIOp>()) &&
+        !matchPattern(extOp0, m_Op<arith::ExtFOp>())) {
+      LLVM_DEBUG(llvm::dbgs() << "no arith.ext body op on in0\n");
+      return false;
+    }
+    Value ext0 = extOp0->getOperand(0);
+    if (!ext0 || !ext0.isa<BlockArgument>()) {
+      return false;
+    }
+  }
+  if (!mul1.isa<BlockArgument>()) {
+    Operation *extOp1 = mul1.getDefiningOp();
+    if (!extOp1 || extOp1->getNumOperands() != 1) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "no arith.ext body op on in1, wrong numOperands\n");
+      return false;
+    }
+    if (!matchPattern(extOp1, m_Op<arith::ExtUIOp>()) &&
+        !matchPattern(extOp1, m_Op<arith::ExtSIOp>()) &&
+        !matchPattern(extOp1, m_Op<arith::ExtFOp>())) {
+      LLVM_DEBUG(llvm::dbgs() << "no arith.ext body op on in1\n");
+      return false;
+    }
+    Value ext1 = extOp1->getOperand(0);
+    if (!ext1 || !ext1.isa<BlockArgument>()) {
+      return false;
+    }
+  }
+  return true;
+}
+
+static SmallVector<int64_t> getInverseTransposePerm(SmallVector<int64_t> perm) {
+  SmallVector<int64_t> inversePerm(perm);
+  for (auto permIdx : llvm::enumerate(perm)) {
+    inversePerm[permIdx.value()] = permIdx.index();
+  }
+  return inversePerm;
+}
+
+static Value transposeTensor(Location loc, PatternRewriter &rewriter,
+                             Value input, SmallVector<int64_t> perm) {
+  if (!perm.size()) {
+    return input;
+  }
+  if (llvm::all_of(llvm::enumerate(perm),
+                   [](auto idx) { return idx.index() == idx.value(); })) {
+    return input;
+  }
+  auto inputType = llvm::cast<RankedTensorType>(input.getType());
+  ArrayRef<int64_t> inputShape = inputType.getShape();
+  SmallVector<int64_t> newInputShape(inputShape);
+  for (auto permIdx : llvm::enumerate(perm)) {
+    newInputShape[permIdx.index()] = inputShape[permIdx.value()];
+  }
```

Comment:
You can use the upstream method `applyPermutation`, which is from https://github.com/llvm/llvm-project/blob/77e88db6b7cd982769a852b0fedfc57b3374d1d9/mlir/include/mlir/Dialect/Utils/IndexingUtils.h#L217-L226

```suggestion
  SmallVector<int64_t> newInputShape = applyPermutation(inputShape, perm);
```